### ISSUE TITLE
Use a cached delegate for method group conversion

### ISSUE THREAD

**Issue by gafter** [dotnet/roslyn#5835]
The method group conversion today always creates a fresh delegate instance.

Instead, for a method group conversion from a static method we should cache it and only create it once, like we do for non-capturing lambda expressions.

[The C# language spec is being modified to permit this]

**Comment by stephentoub** [comment]
WOO HOO!!!

**Comment by danieljohnson2** [comment]
Isn't this a breaking change? I don't think I've ever written code that cares about delegate identity (as opposed to equality), but if it anybody has, this could break it.

Also, it used to be that you got crazy compiler magic if you used the `delegate { ... }` or `foo => bar` syntaxes, and heaven knows what code gets generated then (caching the delegate is the least of it) But the magic happens only if you ask for it.

Now, the rule is to be that the crazy compiler magic happens unless you say `new`. Admittedly, the 'magic' here is really just injecting a small, bounded memory leak, but I think it's still surprising. I was surprised when I found out that `=>` does this sometimes, and I knew about `Expression<T>` already, which is far crazier.

I guess this is mostly harmless, so long as you do not also break the equality semantics of the delegates produced, and so long as you can turn this off with the `new EventHandler(...)` syntax.

**Comment by mburbea** [comment]
Yes please. Can we do something about delegates in generic methods? Currently they are still a much slower caching (if they get cached at all) strategy then for delegates created in non-generic methods

**Comment by Lukazoid** [comment]
:+1: for this, also as per the comment by @mburbea is there any interest in caching for delegates in generic methods? Is it worth creating a separate issue for it.

**Comment by stephentoub** [comment]
> is there any interest in caching for delegates in generic methods? Is it worth creating a separate issue for it.

Roslyn already does this.  For example, with this code:
```C#
using System;

public class Test
{
    public static void Main() { GenericMethod<string>(); }

    private static void GenericMethod<T>()
    {
        InvokeAction(() => Console.WriteLine(typeof(T)));
    }

    private static void InvokeAction(Action action)
    {
        action();
    }
}
```
the code generated for ```GenericMethod<T>``` looks like this:
```
private static void GenericMethod<T>()
{
    InvokeAction(
        <>c__1<T>.<>9__1_0 ?? 
        (<>c__1<T>.<>9__1_0 = new Action(<>c__1<T>.<>9.<GenericMethod>b__1_0)));
}

private sealed class <>c__1<T>
{
    public static readonly Test.<>c__1<T> <>9;
    public static Action <>9__1_0;
    internal void <GenericMethod>b__1_0();
    ...
}
```

**Comment by Lukazoid** [comment]
Oh sweet, I did verify the IL but forgot I wasn't using Roslyn yet at work!

**Comment by diryboy** [comment]
Is the following way of caching expected?
```C#
class MyClass2<K>
{
    class MyClass1
    {
        class MyClass0<V>
        {
            void Test<P>( string[] args )
            {
                    Invoke<P>(Target);
                    Invoke(Cached<K, V, P>.Target ?? (Cached<K, V, P>.Target = new Action<P>(Target)));

                    Invoke<long>(Target);
                    Invoke(Cached<K, V, long>.Target ?? (Cached<K, V, long>.Target = new Action<long>(Target)));
            }

            static void Target<T>( T t ) { }

            static void Invoke<T>( Action<T> a ) { }
        }
    }
}

class Cached<K, V, T>
{
    public static Action<T> Target;
}
```

**Comment by sharwell** [comment]
So what you're saying is the compiler should handle the insanity that was DotNetAnalyzers/StyleCopAnalyzers#1689?

I'm totally in agreement there. :+1:

**Comment by diryboy** [comment]
I created #6642 for the C# part, consider the following code:

```C#
class MyClass2<K>
{
    class MyClass1
    {
        class MyClass0<V>
        {
            void Test<P>( string[] args )
            {
                Invoke<P>(Target);
            }

            static void Target<T>( T t ) { }

            static void Invoke<T>( Action<T> a ) { }
        }
    }
}
```

The compiler will generate a generic class at the top level:
```C#
[CompilerGenerated]
internal static class <>q__MethodGroup0<T0, T1, T2, T3> // Let's call it Cached<...> below
{
	[CompilerGenerated]
	public static T3 <Target>r__Delegate; // Let's call it Delegate below
}
```

And `Invoke<P>(Target)` will be rewritten into something like:
```C#
Invoke(Cached<K, V, P, Action<P>>.Delegate ?? Cached<K, V, P, Action<P>>.Delegate = new Action<P>(Target));
```

The IL looks like:
```IL
// Code size 35 (0x23)
.maxstack 8

IL_0000: nop
IL_0001: ldsfld !3 class '<>q__MethodGroup0`4'<!K, !V, !!P, class [mscorlib]System.Action`1<!!P>>::'<Target>r__Delegate'
IL_0006: dup
IL_0007: brtrue.s IL_001c

IL_0009: pop
IL_000a: ldnull
IL_000b: ldftn void class Test.MyClass2`1/MyClass1/MyClass0`1<!K, !V>::Target<!!P>(!!0)
IL_0011: newobj instance void class [mscorlib]System.Action`1<!!P>::.ctor(object, native int)
IL_0016: dup
IL_0017: stsfld !3 class '<>q__MethodGroup0`4'<!K, !V, !!P, class [mscorlib]System.Action`1<!!P>>::'<Target>r__Delegate'

IL_001c: call void class Test.MyClass2`1/MyClass1/MyClass0`1<!K, !V>::Invoke<!!P>(class [mscorlib]System.Action`1<!!0>)
IL_0021: nop
IL_0022: ret
```

What do you think?

**Comment by mattwar** [comment]
Yes, do this. 👍

**Comment by jcouv** [comment]
We were just discussing this with @CyrusNajmabadi and @cston. We were surprised by the current behavior, as calling with a static method seems like it should not allocate on every invocation.
There is a workaround (pass `() => StaticMethod()` instead of `StaticMethod`) but it is inconvenient.

**Comment by HFadeel** [comment]
Rpslyn can cache ```() => StaticMethod()```  but not ```StaticMethod```
I think should be fixed.

**Comment by marksmeltzer** [comment]
This was brought up in the discussion of #23557....

Any reason this is still pending? I see above that the lang spec requires an update.

**Comment by yaakov-h** [comment]
> [The C# language spec is being modified to permit this]

Is there a corresponding issue on csharplang, or has the spec already been modified?

**Comment by danieljohnson2** [comment]
@kasper3, I don't think they are 'taking it into account'- I think they are going ahead with a breaking change. The spec used to promise a new delegate instance each time you did a method group conversion, and this won't be the case anymore. The new behavior will be less predictable.

That said, while this does demand a spec change, I doubt it will be a real problem. It would take some very strange code to actually *break* due to this. I expect real code will be slightly faster or slower and nothing more.

**Comment by marksmeltzer** [comment]
Has there been any progress on this issue? I see some merged and closed statuses above.

**Comment by sungam3r** [comment]
@stephentoub @gafter  It has been a long time since posting that issue. Has a positive or negative decision been made on the problem described?

**Comment by AqlaSolutions** [comment]
If the quoted code is what is really generated, it looks like it's not thread safe without volatile.
> 
> 
> > is there any interest in caching for delegates in generic methods? Is it worth creating a separate issue for it.
> 
> Roslyn already does this. For example, with this code:
> 
> ```cs
> using System;
> 
> public class Test
> {
>     public static void Main() { GenericMethod<string>(); }
> 
>     private static void GenericMethod<T>()
>     {
>         InvokeAction(() => Console.WriteLine(typeof(T)));
>     }
> 
>     private static void InvokeAction(Action action)
>     {
>         action();
>     }
> }
> ```
> 
> the code generated for `GenericMethod<T>` looks like this:
> 
> ```
> private static void GenericMethod<T>()
> {
>     InvokeAction(
>         <>c__1<T>.<>9__1_0 ?? 
>         (<>c__1<T>.<>9__1_0 = new Action(<>c__1<T>.<>9.<GenericMethod>b__1_0)));
> }
> 
> private sealed class <>c__1<T>
> {
>     public static readonly Test.<>c__1<T> <>9;
>     public static Action <>9__1_0;
>     internal void <GenericMethod>b__1_0();
>     ...
> }
> ```

**Comment by stephentoub** [comment]
> If the quoted code is what is really generated, it looks like it's not thread safe without volatile.

According to ECMA, no, but the runtimes have additional enforcement in place to ensure it is,  in large part because such usage is so common.

**Comment by AqlaSolutions** [comment]
@stephentoub , do you mean write fences after each store or is there something else? Because reading also requires proper handling.

**Comment by weltkante** [comment]
Note that this doesn't need to be 100% threadsafe if its ok to create multiple instances of the delegate when racing (i.e. if the caching is just a performance optimization and not a guarantee to have a unique delegate identity).

In this case its ok to leave away volatile, racing threads may consult the variable in the CPU cache and all create a delegate but only one store wins. Pointer sized stores are guaranteed to be atomic by the runtime (i.e. no tearing), future threads going to the section will see the delegate and not create further ones.

**Comment by stephentoub** [comment]
> do you mean release fences on each store or is there something else? Because reading also requires proper handling.

I mean writes of such object fields will have store release semantics, and reads won't be introduced (which would be the concern on the read side).

**Comment by shadow-cs** [comment]
Any progress on this?

From my point of view since Local functions were added, it might make sense people may start using it in places where they would use Lambdas in the past and unknowingly start to pay the penalty in terms of allocations. The use of Local functions or static methods may feel natural in some situations and the penalty is not easily visible. Also since the recent simplifications in ASP.NET Core even more people may be tempted to use static methods in `MapXxx` calls.

I would also say that recently there has been a lot of fuss about knowing your allocations and making optimizations in all sorts of places so this seems like a logical next step in this direction 😉 (even more so because it is not obvious to the users and may be difficult to explain to less experienced programmers).

**Comment by davidfowl** [comment]
> Also since the recent simplifications in ASP.NET Core even more people may be tempted to use static methods in MapXxx calls.

That allocation is once per application (not per request) so its's not a big deal.

**Comment by shadow-cs** [comment]
Right @davidfowl this is done in startup and then cached internally, I failed to think that far ahead.

**Comment by shadow-cs** [comment]
Any news? Seems to me more and more people stumble upon this.

**Comment by KalleOlaviNiemitalo** [comment]
> Is there a corresponding issue on csharplang, or has the spec already been modified?

The spec has already been modified and published.

ECMA-334 (C# Language Specification) 4th edition (June 2006) disallows caching:

* §13.6 (Method group conversions)

  > The compile-time application of the conversion from `E` to `D` is the same as the compile-time processing of the delegate creation expression `new D(E)` (§14.5.10.3).
* §14.5.10.3 (Delegate creation expressions)

  > The result is a value of type `D`, namely a newly created delegate that refers to the method `M` and target object.

ECMA-334 5th edition (December 2017) allows caching:

* §11.8 (Method group conversions)

  > The conversion is permitted (but not required) to use an existing delegate instance that already contains these references.

Also available [in dotnet/csharpstandard](https://github.com/dotnet/csharpstandard/blob/standard-v5/standard/conversions.md#118-method-group-conversions).

**Comment by RikkiGibson** [comment]
This is resolved by #58288. Expected to ship in 17.2. I believe the caching behavior only occurs when using a preview language version, and probably with C# 11/.NET 7 and up once they are released.

**Comment by jcouv** [comment]
FYI, feature was merged for Visual Studio 17.2 preview 2 as preview feature. Will be released as part of .NET 7 and C# 11 by end of year.

**Comment by davidfowl** [comment]
![ClappingHappyGIF](https://user-images.githubusercontent.com/95136/156237237-3ac2e981-3f00-4862-9b7e-c8c8078b9b68.gif)
